python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+"""
+This uses the expectation maximization algorithm to fit a latent variable
+model for Tada, with two flavors, a ZIP and a standard poisson. The 
+generative formulation is 
+
+p(z) = Dir(alpha_k)
+p(x_ij|z=k) = (ZI)Poisson(Lambda_{jk})
+
+Obviously the ZIP will have a parameter pi for probability of 0.
+
+ZIP does not have an analytic maximum likelihood formula. This is not a 
+problem for EM,  the M step only requires that we be able to maximize
+the likelihood, not that it has an analytic solution. So we just use 
+standard gradient descent with automatic differentiation in torch. This 
+could be naughty if the objective was multimodal, but I strongly doubt that
+this is the case.
+
+Objects
+-------
+MVTadaPoissonEM(K=4,training_options={})
+    This fits a standard poisson latent variable model.
+
+MVTadaZPoissonEM(K=4,training_options={})
+    This fits a standard ZIP latent variable model.
+
+Methods
+-------
+None
+
+"""
+import numpy as np
+from copy import deepcopy
+import scipy.stats as st # type: ignore
+from sklearn.mixture import GaussianMixture # type: ignore
+from tqdm import trange # type: ignore
+
+import torch
+from torch import nn
+from torch.nn import PoissonNLLLoss
+
+
+class MVTadaPoissonEM(object):
+    def __init__(self, K=4, training_options={}):
+        """
+        This is a Product Poisson latent variable model.
+
+        Parameters
+        ----------
+        K : int,default=4
+            The number of clusters, i.e. number of different types of 
+            genes
+
+        training_options : dict,default={}
+            The parameters for the inference scheme
+        """
+        self.K = int(K)
+        self.training_options = self._fill_training_options(training_options)
+
+    def fit(self, X, progress_bar=True, Lamb_init=None, pi_init=None):
+        """
+        Fits a model given count data X.
+
+        Parameters
+        ----------
+        X : np.array-like,shape=(N,p)
+            The count data
+
+        progress_bar : bool,default=True
+            Whether to print a progress bar while fitting
+
+        Lamb_init : np.array-like,shape=(k,p),default=None
+            Initialize the loadings of the model. Defaults to fitting a 
+            GMM on the data and using those.
+
+        pi_init : np.array-like,shape=(k),default=None
+            Initialize the class weights of the model.
+
+        Returns
+        -------
+        self
+        """
+        td = self.training_options
```

Comment:
nit: I think it's fine to use one letter variables for math modeling, but what's `td` short for here?